---
title: "Week 7, Pathway visualization"
author: "Ane Kleiven"
date: "2024-10-17"
output: html_document
---

<H2> Answers to questions <h2/> 

**1. How do PCA and PLS relate? Describe similarities and differences. Write at most 250 words.**


Principal component analysis and PLS are both methods to reduce the dimensionality of different data sets. 

In the PCA method (principal component analysis) you will find the direction in the multidimensional space, where the points (for example different metabolites) are spread the most (the highest variance). 
PCA tries to explain the data as much as possible. The PCA is an unsupervised method, this means that the method doesn’t use any class labels or information about the points to spread the data. 
In PCA you will find the metabolites which explain most of the variation in some direction. 

The PLS method (partial least squares) is a supervised version of the PCA plot. 
In the PLS method, the class labels (information about the points) are used to spread the data. 
PLS searches the direction where the different classes are most spread. If you have a multidimensional space with red and blue points, PLS uses the information of color to spread the data. 


**2. A biologist gives you transcriptomics and metabolomics data for WT and mutant samples. She is particularly interested in the Phenylpropanoid pathway and wants a nice figure in her paper visualizing how change in gene expression relates to change in metabolite abundance between WT and mutant. What do you do? Write at most 250 words.**


To make the figure I would use the method “paintomics” where the genes and metabolites are colored based on their abundance in the pathway. 

Both the genes and the metabolites would have to be coded. 

To code the data, I would use log fold change and decide the average abundance/expression between the mutant and the wildtype. Using the log fold change, you can make a scale showing the difference in abundance/gene expression by color. 
The log fold change will have to be computed for all different genes in the pathway.

If for example a gene is higher expressed in the wildtype, this could get the color blue, and a lower expressed mutant protein could get the color yellow. 
The colors in the map will show which proteins are expressed and not in wildtype and mutant. 

The same can be done for the metabolomics data, where the color represents the abundance of the metabolite. 
A strong color can indicate high abundance, while a weak color can indicate no or low abundance. 
This way, the map can show you which genes affect different metabolites. 
The map can also show alternative paths in the pathway, if some genes are knocked out.  


**3. The biologist in point 2 is happy with the figure, but wants more detailed insight. She has heard about flux balance analysis (FBA), and in a meeting she says: "I have heard that FBA analysis do not use omics data, so I don't think we can use it to analyze my WT/mutant plants". You answer: "Well, you are both right and wrong ...". Finish you reasoning. How can she preform FBA and how can the omics data be used? Write at most 250 words.**


Flux balance analysis is a method for determining the flux of different reactions. 

Determining the flux means to determine how often different reactions in a pathway happens – for example the reaction of ADP to ATP, per time per volume. 

To do this, the biologist has to know the different reactions in the pathway (turn the different metabolites into reactions). 
The reactions are then set up and coded into a stoichiometric matrix, with rows as metabolites and columns as reactions. In the equation set to zero, the different metabolites are multiplied by the fluxes V. The equation is set to zero, assuming everything in the reaction is consumed. 
The reactions are then solved, to find values for the V’s (fluxes) that maximizes the biomasses.

To include omics data in the FBA, you assign constraints based on the omics data. If you know that some protein in the pathway is low-abundance, you assign this constraint to the FBA. This means forcing some fluxes to be high or low, based on what the omics data tells us.

Using omics data makes the experiments more realistic, because you can change the constraints based on what you’re looking at. A protein in the brain tissue might not be abundant in the liver.  

Without including omics data, FBA will be very general. In this case you assume that all proteins are abundant in all types of cells at all times, and this is certainly not the case. 



<H2> Pathview <h2/> 


<h3> Install pathview package <h3/> 

```{r, message=FALSE, warning=FALSE}

if (!requireNamespace("BiocManager", quietly = TRUE))
  install.packages("BiocManager")
BiocManager::install("pathview")

```


<h3> Example <h3/> 

```{r, message=FALSE,warning=FALSE}
library(pathview)
data(gse16873.d)

```

<h4> Pathway using one sample in the data set <h4/> 

```{r, message=FALSE, warning=FALSE}

# For human (hsa) pathway cell cycle (04110), plot expression values from the first sample (column 1)
# node.sum set to default. 
pv.out <- pathview(gene.data = gse16873.d[, 1], pathway.id = "04110", species = "hsa", out.suffix = "gse16873.1")

```

```{r, message=FALSE, warning=FALSE} 
# sample one 
knitr::include_graphics("C:/Users/anekl/OneDrive - Norwegian University of Life Sciences/BIN315/hsa04110.gse16873.1.png")

```
 
<h4> Pathway using all the samples in the data set <h4/> 

```{r, message=FALSE, warning=FALSE}
# all samples. node.sum set to default 

pv.out_all <- pathview(gene.data = gse16873.d, pathway.id = "04110", species = "hsa", out.suffix = "gse16873")

```

```{r, message=FALSE, warning=FALSE}
# all samples 

knitr::include_graphics("C:/Users/anekl/OneDrive - Norwegian University of Life Sciences/BIN315/hsa04110.gse16873.multi.png")
```
  
<h3> Gene to node data mapping <h3/> 

<h4> node.sum set to default <h4/>

```{r, message=FALSE, warning=FALSE}
# node.sum set to default for sample 1 (see mol.data for node values) 

head(pv.out$plot.data.gene) 
```
  
<h4> node.sum set to "mean" <h4/>

```{r, message=FALSE, warning=FALSE}
# sample one using node.sum = "mean" 

pv.out_mean <- pathview(gene.data = gse16873.d[,1], pathway.id = "04110", species = "hsa", out.suffix = "gse16873.mean", node.sum = "mean")
```

```{r, message=FALSE, warning=FALSE}
# sample one using node.sum = "mean" 

head(pv.out_mean$plot.data.gene) 
```

**Check if nodes that have several genes now contain the mean value** 

By comparing the two tables above, you see that the mol.data has changed. The value is equal for the nodes that only contains one gene, the values is changed for the nodes containing several genes. Here the values are lower, because the mean value will be lower than the sum of numbers. 

For example: Genes with label MCM2 have a node value of -0.42 in the node.sum as default. Since this label contains 6 genes, the node value is -0.07 using node.sum = "mean" (-0.42/6 = -0.07). 


```{r, include=FALSE}
# generate random metabolomic data 

sim.cpd.data = sim.mol.data(mol.type = "cpd", nmol = 3000)

pathview(gene.data = gse16873.d[,1], cpd.data = sim.cpd.data, pathway.id = "00640", species = "hsa", out.suffix = "gse16873.mbl")
```


```{r, include=FALSE}
knitr::include_graphics("C:/Users/anekl/OneDrive - Norwegian University of Life Sciences/BIN315/hsa00640.gse16873.mbl.png")
```

<H2> Example using autism gene expression data <h2/> 

<h3> KEGG pathway gene enrichment <h3/> 

<h4> Load the previous differential expression results from lab 3 <h4/>

```{r, message=FALSE, warning=FALSE}
load("autism_DEA.Rdata")

head(res_0.05FC)
```

<h4> Visualize the fold change in gene expression levels between autism and control on the selected enriched pathway <h4/>

```{r, message=FALSE, warning=FALSE}

library(dplyr) 
library(tibble) 
# make a data frame with the log2FoldChange 

log2fc_genes <- res_0.05FC %>% column_to_rownames(var = "Gene") %>% select(log2FoldChange) %>% as.data.frame()

```

```{r, message=FALSE, warning=FALSE}
# chosen pathway: Synaptic vesicle cycle, HSA04721  

synaptic <- pathview(gene.data = log2fc_genes, pathway.id = "04721", species = "hsa", out.suffix = "synaptic", gene.idtype = "ENSEMBL")
```

```{r}
head(synaptic$plot.data.gene)
```

```{r, message=FALSE, warning=FALSE}
knitr::include_graphics("C:/Users/anekl/OneDrive - Norwegian University of Life Sciences/BIN315/hsa04721.synaptic.png")
```
  
<h4> Experiment with some settings <h4/>

```{r, message=FALSE, warning=FALSE}

# setting the node.sum to mean 
synaptic_mean <- pathview(gene.data = log2fc_genes, pathway.id = "04721", species = "hsa", out.suffix = "synaptic mean", gene.idtype = "ENSEMBL", node.sum = "mean")
```

```{r, message=FALSE, warning=FALSE}
head(synaptic_mean$plot.data.gene)
```
In the table above, we can see that the node value has changed for the nodes containing several genes. 

```{r, message=FALSE, warning=FALSE}
# Change the color scaling using limit: 

synaptic_limit <- pathview(gene.data = log2fc_genes, pathway.id = "04721", species = "hsa", out.suffix = "synaptic_limit", gene.idtype = "ENSEMBL", node.sum = "mean", limit = list(gene = 3))
                                 
```

```{r, message=FALSE, warning=FALSE}
knitr::include_graphics("C:/Users/anekl/OneDrive - Norwegian University of Life Sciences/BIN315/hsa04721.synaptic_limit.png")
```
 
<h4> Visualize the signed log10 p-values in the pathway <h4/>

```{r, message=FALSE, warning=FALSE}
# make a data frame with the signed log10 p-values 
library(dplyr) 
library(tibble)

signed_pvalue <- res_0.05FC %>% mutate(log10_pvalue = -log10(pvalue), sign_FC = sign(log2FoldChange),         signed_p = log10_pvalue * sign_FC) %>% column_to_rownames(var = "Gene") %>% select(signed_p) %>% as.data.frame()                             

head(signed_pvalue)
```
  
<H4> Using limit set to 17 <h4/>

```{r, warning=FALSE, message=FALSE}
# synaptic pathway using signed p-values
# limit set to 17, covering the full range of p-values

synaptic_pvalue <- pathview(gene.data = signed_pvalue, pathway.id = "04721", species = "hsa", out.suffix = "synaptic_pvalue", gene.idtype = "ENSEMBL", node.sum = "mean", limit = list(gene = 17))
                                 
```

```{r, message=FALSE, warning=FALSE}
head(synaptic_pvalue$plot.data.gene)
```

```{r, message=FALSE, warning=FALSE}
knitr::include_graphics("C:/Users/anekl/OneDrive - Norwegian University of Life Sciences/BIN315/hsa04721.synaptic_pvalue.png")
```
<H4> Using limit set to 5 <h4/>

```{r, message=FALSE, warning=FALSE}
# synaptic pathway using signed p-values
# limit set to 5

synaptic_pvalue5 <- pathview(gene.data = signed_pvalue, pathway.id = "04721", species = "hsa", out.suffix = "synaptic_pvalue5", gene.idtype = "ENSEMBL", node.sum = "mean", limit = list(gene = 5))
```

```{r, message=FALSE, warning=FALSE}
knitr::include_graphics("C:/Users/anekl/OneDrive - Norwegian University of Life Sciences/BIN315/hsa04721.synaptic_pvalue5.png")
```

<h4>Based on the painted pathway, can you put together a biological story? Does the results make biological sense?<h4/>

In the paper, the lowly expressed genes in autism patients, were enriched for gene ontology categories related to synaptic functions. This is why I chose the "synaptic vesicle cycle". 
In all the pathway maps I made, several genes in the pathway are downregulated (the green colour) for autism patients. 
After searching the web, I found research showing that several of these genes are associated with autism. 
The node syntaxin (in the pathway), contains several genes. One of them is the STX1. Researchers have found several variants for this gene in autism patients (https://www.thelancet.com/article/S2352-3964(15)00030-4/fulltext). 
The VAMP2 gene is also associated wirh autism (https://www.nature.com/articles/s41380-022-01822-1). 

My findings indicate that synaptic function is affected in autism patients. 


<H3> Example with the GABAergic synapse pathway <h3/>

<h4> Using log2FC-values <h4/> 

```{r, message=FALSE, warning=FALSE}

# chosen pathway: GABAergic synapse, pathway ID: HSA04727
GABAergic_synapse <- pathview(gene.data = log2fc_genes, pathway.id = "04727", species = "hsa", out.suffix = "GABAergic", gene.idtype = "ENSEMBL", node.sum = "mean", limit = list(gene = 2))
```

```{r, message=FALSE, warning=FALSE}
knitr::include_graphics("C:/Users/anekl/OneDrive - Norwegian University of Life Sciences/BIN315/hsa04727.GABAergic.png")
```
 
<h4> Using signed p-values <h4/> 

```{r, message=FALSE, warning=FALSE}

# chosen pathway: GABAergic synapse, pathway ID: HSA04727
GABAergic_p <- pathview(gene.data = signed_pvalue, pathway.id = "04727", species = "hsa", out.suffix = "GABAergic_p", gene.idtype = "ENSEMBL", node.sum = "mean", limit = list(gene = 5))
```

```{r, message=FALSE, warning=FALSE}
knitr::include_graphics("C:/Users/anekl/OneDrive - Norwegian University of Life Sciences/BIN315/hsa04727.GABAergic_p.png")
```
  
<H3> Example with the cholinergic synapse pathway <h3/>

<h4> Using log2FC-values <h4/> 

```{r, message=FALSE, warning=FALSE}

# chosen pathway: cholinergic synapse, pathway ID: HSA04725
cholinergic_synapse <- pathview(gene.data = log2fc_genes, pathway.id = "04725", species = "hsa", out.suffix = "cholinergic", gene.idtype = "ENSEMBL", node.sum = "mean", limit = list(gene=2))
```

```{r, message=FALSE, warning=FALSE}
knitr::include_graphics("C:/Users/anekl/OneDrive - Norwegian University of Life Sciences/BIN315/hsa04725.cholinergic.png")
```
 
<h4> Using signed p-values <h4/> 

```{r, message=FALSE, warning=FALSE}
# chosen pathway: cholinergic synapse, pathway ID: HSA04725
cholinergic_p <- pathview(gene.data = signed_pvalue, pathway.id = "04725", species = "hsa", out.suffix = "cholinergic_p", gene.idtype = "ENSEMBL", node.sum = "mean", limit = list(gene=5))
```

```{r, message=FALSE, warning=FALSE}
knitr::include_graphics("C:/Users/anekl/OneDrive - Norwegian University of Life Sciences/BIN315/hsa04725.cholinergic_p.png")
```
<br><br>  
