---
title: "Project BIN315"
author: "Ane Kleiven"
date: "2024-11-02"
output: 
  html_document: 
    code_folding: hide 
---

## Answers to questions 

The omnigenics model suggests that heritability for traits is driven by many variants with small effects. The model has found that traits often have core-genes responsible for the strongest heritability. Much of the heritability also comes from non-core genes. Non-core genes often act through cellular networks. 

The omnigenics model differs from the traditional genetic models (mendelian and polygenic). The Mendelian model suggest that traits are decided by one specific gene. The polygenic model suggest that traits are decided by multiple genes with small, additive effects. The traditional models only explain parts of the heritability, while the omnigenics model explains the missing heritability (at least some of it). 

GWAS have found many variants with small effects in diseases like Crohn’s, Schizophrenia and Rheumatoid arthritis. The effects have very low effect alone, no single variant is dominant alone. 
The variants with largest effects are mostly located in genes or pathways directly related to the disease or trait, but variants can be spread across the genome. 
GWAS found that many of the variants are in non-coding regions in the genome, often far from the core-genes. Many of the variants found, are in regions responsible for gene regulation – enhancers, promoters etc.  

The omnigenics model is a complex model. It explains more of the heritability of traits and gives us the opportunity to understand more of the genome. I believe this model will be important in the future, for explaining diseases and other traits. 

I think this model is more correct than traditional methods used in the past. The technology has developed a lot. DNA-sequencing and bioinformatics tools, gives us the opportunity to go deeper into the genome. 


<H2>Coding exercises<H2/>

Exercise 1.1 Display the three first rows of the mtcars dataset

```{r}
data("mtcars")

mtcars[c(1:3),c(1:11)]
```

Exercise 1.2 a) Cars with 5 gears

```{r}

cars5gears <- subset(mtcars, gear == 5)
cars5gears

```

Exercise 1.2 b) Calculate how many cars that have 5 gear

```{r}

mtcars$gear == 5
sum(mtcars$gear ==5)
```

Exercice 1.2 c) Extract only the names of the cars that have 5 gears, as a character vector.

```{r}

rownames(mtcars[mtcars$gear == 5, ])
```

Exercise 1.2 d) Display the subset of the table with only cars that have 4 gears and at least 100 hp

```{r echo=TRUE}
cars4gears100hp <- subset(mtcars, gear == 4 & hp > 100)
cars4gears100hp
```

Exercise 1.3 a) Display the subset of the table with only Toyotas (hint: grepl()).

```{r}
Toyota <- mtcars[grepl("Toyota", rownames(mtcars)), ]
Toyota
```

Exercise 1.3 b) Create a subset of the table with only Mercedes but only include the two columns mpg and hp

```{r}
Mercedes <- mtcars[grepl("Merc", rownames(mtcars)), ]
Mercedes
Mercedes[, c(1,4)]
```

Exercise 1.4 Use sample() to display 5 random rows from the mtcars table.

```{r}
nrow(mtcars)
mtcars[sample(nrow(mtcars),3), ]

```

Exercise 1.5 Use the order function to display the 5 least fuel efficient cars (lowest mpg).

```{r}
lowest_mpg <- mtcars[order(mtcars$mpg), ]
lowest_mpg[1:5,]
```

Exercise 2.1 a) Calculate the square root of 144 and then round the result to 1 decimal place using the pipe operator.

```{r, message = FALSE}
library(dplyr)
sqrt(144) %>% round(digits = 1)
```

Exercise 2.2 a) Select the name, height, and mass columns from the starwars dataset

```{r}
library(dplyr)
starwars %>% select(name, height, mass) %>% print
```

Exercise 2.2 b) Arrange the starwars dataset by mass in descending order.

```{r}
starwars %>% arrange(desc(mass)) %>% print()
```

Exercise 2.2 c) Filter the starwars dataset to include only characters taller than 180 cm.

```{r}
library(dplyr)
starwars %>% dplyr::filter(height > 180) %>% print()
```

Exercise 2.2 d) Create a new variable in the starwars dataset called bmi using the formula bmi = mass / (height/100)^2, and select the name, height, mass and bmi columns.

```{r}
starwars %>% mutate(BMI = as.numeric(mass / (height / 100)^2)) %>% select(name, height, mass, BMI) %>% print
```

Exercise 2.2 e) Group the starwars dataset by species and calculate the average height for each species

```{r}
starwars %>% group_by(species) %>% summarise(mean(height, na.rm = TRUE)) %>% print
```

Exercise 2.3 a) 
```{r, message=FALSE}
library(tidyverse)
longstarwars <- starwars %>% pivot_longer(cols= c(height,mass,birth_year), names_to = "attribute", values_to = "measurement") %>% print
```

Exercise 2.3 b) Use pivot_wider() to convert the starwars_long dataset back to its original wide format, spreading the measurement_type values across columns.

```{r}
longstarwars %>% pivot_wider(names_from = "attribute", values_from = "measurement")  %>% print
```

Exercise 2.3 c) Separate the name column into first_name and last_name based on the first space character. This will split the names of the characters into two distinct columns.

```{r}
starwars %>% separate(name, into = c("first name","last name"),sep = " ", extra = "merge", fill = "right") %>% print
```

Exercise 2.4 Write the starwars dataset to a new CSV file called starwars_output.csv.

```{r}
starwars[] <- lapply(starwars, function(x) {if (is.list(x)) {sapply(x, paste, collapse = ", ")} else {x}})

write.csv(starwars, file = "C:/Users/anekl/OneDrive - Norwegian University of Life Sciences/starwars_output.csv", row.names = FALSE) 
```

Exercise 2.5 
Using the starwars dataset, calculate the average mass for each species, and then filter to include only species with an average mass greater than 100. Finally, write this filtered dataset to a CSV file.

```{r}
library(dplyr)
data("starwars")

starwarsmass100 <- starwars %>% group_by(species) %>% summarise(meanmass = mean(mass, na.rm = TRUE)) %>% dplyr::filter(meanmass > 100) %>% print

```
```{r}
write.csv(starwarsmass100, file = "C:/Users/anekl/OneDrive - Norwegian University of Life Sciences/starwarsmass100_output.csv", row.names = FALSE)
```


Exercise 3.1 Recreate the scatter plot with points colored based on the number of gears using ggplot2

```{r}
library(ggplot2)
data("mtcars")

mtcars %>% ggplot(aes(x = disp, y = mpg, color = factor(gear))) + geom_point(size = 4) + labs(title = "Highway Fuel Efficiency and Engine Displacement", x = "Engine Displacement (Liters)", y = "Highway Miles per Gallon")
```

Exercise 3.2 Recreate the plot using ggplot2

```{r, message=FALSE}
library(ggplot2)
library(dplyr)

vehicles <- mtcars %>% mutate(vehicle_class = case_when(gear == 2 & wt < 3.0 ~ "2-Seater", wt < 2.5 ~ "Subcompact",  wt >= 2.5 & wt < 3.0 ~ "Compact",   wt >= 3.0 & wt < 4.0 ~ "Midsize", wt >= 4.0 & disp > 300 ~ "SUV", wt >= 4.0 & disp <= 300 ~ "Pickup",  wt >= 3.0 & mpg > 15 ~ "Minivan", hp > 230 & wt < 3.5 ~ "Sports Car",   TRUE ~ "Other")) %>% print()

mtcars_modified <- vehicles %>% mutate(drv = case_when(cyl == 4 & wt < 3.0 ~ "FWD", cyl == 6 & wt >= 3.0 & wt < 4.0 ~ "RWD", cyl == 8 ~ "AWD", TRUE ~ "Unknown"))

mtcars_modified %>% ggplot(aes(x = disp, y = mpg, color = vehicle_class, size = factor(cyl))) + geom_point() + facet_wrap(~drv) + labs(title = "Highway Fuel Efficiency and Engine Displacement", subtitle = "Faceted by Drive Type, Colored by Vehicle Class, Sized by Cylinders", x = "Engine Displacement (Liters)", y = "Highway Miles per Gallon", color = "Vehicle Class", size = "Number of Cylinders") %>% print()

```


