---
title: "Week 2 Transcriptomics"
author: "Ane Kleiven"
date: "2024-09-12"
output: html_document
---

<H2>Answers to questions<H2/>

**1. Why do we log-transform data?**

We log-transform to make the data more readable and stretched out. The log value spreads the values more evenly across the value range, this will often give a normal distribution. 
On a linear scale, most of the values will be close to zero with outliers with high values.
Log transformation makes the data more symmetrical. 


**Many scaling methods (i.e. normalisation for library size in RNA-Seq data) assume that most genes are not differentially expressed. Why is this a necessary assumption for methods such as DESeq2 and edgeR/TMM?**

An assumption used in the methods is that the values are not differentially expressed. This is done to make the gene expression more similar across samples. Genes with low expression will be upscaled, genes with high expression will be downscaled. 
The assumption is necessary to make the methods work. If the majority of genes are not differentially expressed, the method will erase the systematical differences and keep the actual differences in the minority of differentially expressed genes. 
Makes the non-differentially expressed genes more similar. 

If most genes actually are different, the method will erase the differences. 


**3. Explain the intuition behind the EM-algorithm for estimating gene expression from multi-mapping reads. What is the problem that this algorithm is trying to solve and how does it solve it?**

EM-algorithm is a method for assigning multireads probabilistically.The problem the algorithm is trying to solve, are the multi-mapping-reads, that could fit several isoforms. The algorithm predicts which read belongs to which isoform. 

The algorithm start by looking at the fraction of reads mapping to isoform 1 and 2 (theta 1 and 2)
When we have an estimate of these thetas, we can find an estimate of tau (fraction of transcripts (mRNA) belonging to isoform i (the gene). 
The algorithm is iterative, and is improved by adding the multireads to the theta. This will give us a new tau. These steps are repeated until the result no longer changes. The end result is the probable distribution of the multi-reads. 


<H2>Installing packages<H2/>

```{r cars, message = FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("DESeq2")

```


<H2>Loading the count data<H2/>

```{r, message=FALSE}
library(tidyverse)
library(dplyr)
library(readr)
read_tsv("AspWood/K1-01_sortmerna_trimmomatic/quant.sf", show_col_types = FALSE) %>% head(10)
```

Potra2n... are genes from the tree species Populus tremula (Aspen). For example, Potra2n765s36714.1 is a gene coding for the enzyme Chitooligosaccharide oxidase. The number 765 stands for scaffold765, indicating the region of the genome where the transcript is located. 

TPM (Transcripts Per Million) is a normalization method in RNA-Seq to account for the length of the transcript and the total number of reads. TPM makes it possible to compare gene expression across genes or isoforms within the same sample. 
The sum of TPM's in one sample are the same for all samples. This gives us the opportunity to compare gene expression between samples. 


```{r, message=FALSE}
library(readr)
library(dplyr)

data_Potra <- read_tsv("AspWood/K1-01_sortmerna_trimmomatic/quant.sf", show_col_types = FALSE) %>% head(10)

data_Potra %>% summarise(Total_TPM = sum(TPM))
```
The last column NumReads is the estimated number of reads mapped to the transcript. These numbers are not integers because they are numbers with decimals. In R, integers are whole numbers, while numerical values are numbers with decimals. 

The sum of the NumReads column are the total number of reads in this sample. 

```{r}
data_Potra %>% summarise(Tot_NumReads = sum(NumReads))
```

```{r, message=FALSE, echo=FALSE}

#making a counts table 

library(readr)
library(dplyr)
library(tibble)

samples <- list.files("AspWood")

counts <- tibble()
for (sample in samples) {
  file <- paste0("AspWood/", sample, "/quant.sf")
  
  sample.trimmed <- gsub("_sortmerna_trimmomatic", "", sample)
  
  c <- read_tsv(file) %>%
    select(Name, NumReads) %>%
    rename(Genes = Name, !!sym(sample.trimmed) := NumReads)
  
  if (sample == samples[1]) {
    counts <- c 
  } else {
    counts <- cbind(counts, c %>% select(-Genes))
  }
}

dim(counts)
```


```{r, echo=FALSE, results='hide'}
#the number of reads mapped to a specific transcript in a specific sample

library(readr)
library(dplyr)

counts %>% select(Genes, `K1-02`) %>% filter(Genes == "Potra2n765s36714.1") 
```


```{r, message=FALSE, echo=FALSE, results='hide'}
#Sum the counts for each loci method 1. 

library(readr)
library(dplyr)

counts <- counts %>% pivot_longer(cols = -Genes, names_to = "Samples", values_to = "Expression") %>% separate(Genes, into = c("Genes"), sep = "\\.", extra = "drop") %>% group_by(Genes, Samples) %>% summarise(Expression = sum(Expression)) %>% pivot_wider(names_from = "Samples", values_from = "Expression")
```


```{r, echo=FALSE, results='hide'}

#Sum the counts for each loci method 2. 

library(readr)
library(dplyr)

counts <- counts %>% separate(Genes, into = c("Genes"), sep = "\\.", extra = "drop") %>%  group_by(Genes) %>% summarise_if(is.numeric, sum)

dim(counts)
```


```{r, echo=FALSE, results = "hide"}
#Storing a long version of the counts table 

library(readr)
library(dplyr)

counts.long <- counts %>% pivot_longer(cols = -Genes, names_to = "Samples", values_to = "Expression")

dim(counts.long)

```



<H2>Exploring the count data and the log-transform<H2/>

Make a boxplot of the counts 

```{r, message=FALSE}
library(readr)
library(dplyr)

ggplot(counts.long, aes(x = Samples, y = Expression)) + geom_boxplot() + ggtitle("Boxplot of Gene Expression by Sample") + ylab("Expression") + xlab ("Samples") + theme(axis.text.x = element_text(size = 6, angle = 30, hjust = 1), axis.text.y = element_text(size = 6))

```
**Removing the outliers:** 

```{r, message=FALSE}
library(readr)
library(dplyr)

ggplot(counts.long, aes(x = Samples, y = Expression)) + geom_boxplot(outliers = FALSE) + ggtitle("Boxplot of Gene Expression by Sample") + ylab("Expression") + xlab ("Samples") + theme(axis.text.x = element_text(size = 6, angle = 30, hjust = 1), axis.text.y = element_text(size = 6))

```
The boxplot above show the variation in gene expression between different samples. The white boxes includes 50% of the counted gene expression. The black horisontal line is the median value of the gene expression. The vertical lines is the total spread within the group. The end of the vertical line is the maximum value within the given sample. 

The boxplot above show that sample K1-05 have the highest gene expression (this is an outlier in the data set) - both in median value and maximum value. The gene expression between the other samples differs. The most equal groups (K1-15 - K1-21) have a median value close to zero, indicating that no genes are expressed in these samples.


**A new boxplot where the expression values have been log2-transformed**


Here we use log2-transformation to make the data more readable and stretched out. The log-transformation can make the data more normally distributed. Because of the outliers, many values will look like they are close to zero on a linear scale. We use Log2 (expression + 1) because log(0) is not defined. 

```{r, message=FALSE}

library(dplyr)

log_expression <-  log2(counts.long$Expression + 1)

ggplot(counts.long, aes(x = Samples, y = log_expression)) + geom_boxplot(outliers = FALSE) + ggtitle("Boxplot of Gene Expression by Sample") + ylab("Expression Log2") + xlab ("Samples") + theme(axis.text.x = element_text(size = 6, angle = 30, hjust = 1), axis.text.y = element_text(size = 6))
```
Reading log-transformed values means that an increase from 5 to 6 on the log2-scale doubles on the linear scale (2^5 = 32 means 64 on the linear scale). If we had used log10 instead, the increase from 5 to 6 on the log2-scale would mean 10x on the linear scale. 

Based on the histograms showing the linear and the log-scaled version, the main effect of the log-transformation is to make the data more spread and readable. The log2-transformed histogram shows that the data are more normally distributed. The zero-column is quite high. This tells us that a lot of the samples have zero gene expression. 



<H2>Normalization<H2/>

```{r, message=FALSE,results='hide'}
#library(DESeq2)

counts.mat <- counts %>% column_to_rownames(var = "Genes") %>% as.matrix() %>% round()

dds <- DESeq2::DESeqDataSetFromMatrix(countData = counts.mat,colData = data.frame(conditions = as.factor(colnames(counts.mat))),design = ~ conditions)
```

The count table after normalization, will show that gene expression between the different samples are almost the same. The normalization step will account for different library sizes. 

```{r, message=FALSE}
dds <- DESeq2::estimateSizeFactors(dds)
DESeq2::counts(dds, normalized = TRUE) %>% as.data.frame() %>% rownames_to_column(var = "Genes") %>%  
pivot_longer(cols = -Genes, names_to = "Samples", values_to = "Expression") %>% mutate(Expression = log2(Expression + 1)) %>% ggplot(aes(x = Samples, y = Expression)) + geom_boxplot() + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
```
**Size factors** 

```{r}
size_factors <- DESeq2::sizeFactors(dds)
print(size_factors)
```
The size factors are the factors used to account for different library sizes (sequence depth) between samples.
A size dactor > 1 means that the sample has a larger library size compared to the other samples. 
A size factor < 1 means that the sample has a smaller library size compared to the other samples. 

The expression in the boxplots are much more equal after normalization. 

K1-05 has the highest expression in the boxplot prior to normalization, this has a size factor of 3.49, meaning that the sample has a smaller library size compared to the other samples - giving a higher expression value before normalization. 



<H2>Variance stabilizing transformation (VST)<H2/>

```{r, message=FALSE}
library(dplyr)
vst <- DESeq2::varianceStabilizingTransformation(counts.mat)
vst <- vst - min(vst)
```

**Boxplot of the VST-normalized values for each sample:**

```{r}
library(dplyr)

as.data.frame(vst) %>% rownames_to_column(var = "Genes") %>%  
pivot_longer(cols = -Genes, names_to = "Samples", values_to = "Expression") %>% mutate(Expression = log2(Expression + 1)) %>% ggplot(aes(x = Samples, y = Expression)) + geom_boxplot() + ggtitle("Boxplot of Gene Expression by Sample") + ylab("Expression") + xlab("Samples") + theme(axis.text.x = element_text(size = 6, angle = 30, hjust = 1), axis.text.y = element_text(size = 6))
```
As described in the AspWood paper, the VST ensures that the expression values are comparable across samples. This fits good with the VST-boxplot above, where all the samples has an expression value ~ 3. 
The expression values are log2-transformed and adjusted for differences in library size (sequencing depth)



<H2>Removing lowly expressed genes<H2/>


The number of genes not expressed in our data: 

```{r,message=FALSE}
library(dplyr)

vst_df <- as.data.frame(vst) %>% rownames_to_column(var = "Genes") 
unexpressed_genes <- vst_df %>% dplyr::filter(if_all(-Genes, ~ .x == 0))
nrow(unexpressed_genes)

```
How many genes have less than 10 mapped reads in total in our data? Remove these genes from vst

```{r, message = FALSE}
library(dplyr)
sum_reads <- vst_df %>% mutate(TotalReads = rowSums(select(., -Genes))) 
lowexpression_genes <- sum_reads %>% dplyr::filter(TotalReads<10)
vst_corrected <- sum_reads %>% dplyr::filter(TotalReads >= 10) %>% select(-TotalReads)

nrow(lowexpression_genes)
nrow(vst_corrected)

```
There are 10890 genes with less than 10 reads. 

After filtering this out, we are left with 26185 genes. 



<H2>Save the data for later<H2/>

```{r}
save(dds, vst, file="AspWood_normalized.RData")
```



<H2>Finally - have a look at the data!<H2/>

```{r, message=FALSE}
library(dplyr)
library(tidyr)
library(ggplot2)


Genes_plot <- c("Potra2n4c9149", "Potra2n16c30563", "Potra2n1c2087","Potra2n4c8952", "Potra2n689s36475")

vst_long <- vst_corrected %>% pivot_longer(cols = -Genes, names_to = "Samples", values_to = "Expression") %>% filter(Genes %in% Genes_plot) 
ggplot(vst_long, aes(x = Samples, y = Expression, color = Genes, group = Genes)) + geom_line() + labs(title = "Gene expression", ylab = "Expression", xlab = "Samples") +  scale_color_viridis_d() + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))
```
SUS6: Potra2n4c9149
CDC2: Potra2n16c30563
EXPA1: Potra2n1c2087
CesA8: Potra2n4c8952
BFN1: Potra2n689s36475

The expression of the genes are not shown for all 5 in the AspWood paper, but the plot for CesA8 looks very much alike the plot I made for this gene. 

